### PR TITLE
[BL-339]Search api

### PR DESCRIPTION
**Pull request by giyyanan**

### CODE REVIEW

File: db/sql/sql.go [hunk 1]
```diff
@@ -1215,6 +1217,12 @@ func AddFilterToQuery(s *schema.Schema, q sq.SelectBuilder, filter map[string]in
 			column = quote(key)
 		}
 
+		substr,ok := value.(string)
+		if ok && strings.HasPrefix(substr,"~"){
+		 	q = q.Where(sq.Like{column: buildLikeValues(substr[1:])})
+		 	continue
+		}
```
Comment (morrisson) [inline db/sql/sql.go]: I got your idea but see my comment below.

File: server/resources/resource_management.go [hunk 2]
```diff
@@ -316,6 +316,23 @@ func GetMultipleResources(context middleware.Context, dataStore db.DB, resourceS
 	currCond := policy.GetCurrentResourceCondition()
 
 	filter := FilterFromQueryParameter(resourceSchema, queryParameters)
+	if search_filter,ok := queryParameters["search_field"]; ok && resourceSchema.IsSubstringSeachEnabled(){
+			for _,column := range search_filter{
+				if search_value,ok := queryParameters[column]; ok {
+					delete(filter, column)
+					searchValue := strings.Builder{}
+					searchValue.WriteString("~")
```
Comment (morrisson) [inline server/resources/resource_management.go]: I think tweaking original input with magic prefix tilde is not good idea. It makes hard for other developers to understand how it works, and may occur some unexpected errors (e.g., users may use tilde as name field, description field and so on...). 

We should manage what properties are used for substr match in other fields. Actually, I like @michalzurawski 's idea, which is discussed offline, that enhance `paginator` below.

File: server/resources/resource_management.go [hunk 3]
```diff
@@ -316,6 +316,23 @@ func GetMultipleResources(context middleware.Context, dataStore db.DB, resourceS
...
+					filter[column] = searchValue.String()
+				} else{
+					err := fmt.Errorf("search value for `%s` not available in URL ", column)
+					return ResourceError{err, err.Error(), WrongQuery}
+				}
+			}
+			log.Info("substring search is enabled")
+	}
+
+	extendFilterByTenantAndDomain(resourceSchema, filter, schema.ActionRead, currCond, auth)
```
Comment (morrisson) [inline server/resources/resource_management.go]: `extendFilterByTenantAndDomain` is doubled.

File: db/sql/sql_test.go [hunk 4]
```diff
@@ -411,6 +411,29 @@ var _ = Describe("Sql", func() {
...
+			})
+			It("should create like query with two parameter", func() {
+				filter := map[string]interface{}{"test_string": "~123", "test_number": "~42"}
+
+				res, err := AddFilterToQuery(testSchema, query, filter, false)
+
+				Expect(err).ToNot(HaveOccurred())
+				resSql, param, err := res.ToSql()
+				Expect(err).ToNot(HaveOccurred())
+				checkAndStatement(resSql, "`test_string` LIKE ?", "`test_number` LIKE ?", param, []interface{}{"%42%", "%123%"})
```
Comment (cierpuchaw) [inline db/sql/sql_test.go]: `checkAndStatement` does not check order, only presence. But please be consistent and reverse what you pass as `expectedParams`.

File: server/resources/resource_management.go (hunk 2 above)
Comment (cierpuchaw) [reply server/resources/resource_management.go]: +1 about `~` being problematic. Can you guys paste Michal's idea here?

File: server/server_test.go [hunk 5]
```diff
@@ -1273,6 +1273,39 @@ var _ = Describe("Server package test", func() {
...
+		})
+
+		AfterEach(func() {
+			testURL("DELETE", getNetworkSingularURL("red"), adminTokenID, nil, http.StatusNoContent)
+			testURL("DELETE", getNetworkSingularURL("red1"), adminTokenID, nil, http.StatusNoContent)
+			testURL("DELETE", getNetworkSingularURL("red2"), adminTokenID, nil, http.StatusNoContent)
+			testURL("DELETE", getNetworkSingularURL("red3"), adminTokenID, nil, http.StatusNoContent)
+		})
+
+		DescribeTable("match all rows by substring search", func(query string) {
```
Comment (cierpuchaw) [inline server/server_test.go]: If you move `func(query string...` to next line, goimports/gofmt will format it a bit better. This is of course subjective.

File: server/server_test.go [hunk 6]
```diff
@@ -1273,6 +1273,39 @@ var _ = Describe("Server package test", func() {
...
+		},
+			Entry("empty query", ""),
+			Entry("long prefix", "?id=networkred&search_field=id"),
+			Entry("substring at the start", "?id=net&search_field=id"),
+			Entry("substring at the middle", "?id=work&search_field=id"),
+			Entry("substring at the last", "?id=red&search_field=id"),
+			Entry("search on one parameter", "?description=red&search_field=description"),
+			Entry("more than one field to search","?id=red&search_field=id&description=red&search_field=description"),
+		)
+	})
```
Comment (cierpuchaw) [inline server/server_test.go]: Can you also add some tests to show that we perform exact matches by default?

Comment (cierpuchaw) [reply server/server_test.go]: `search_field=` does not really explain what it does.

File: server/resources/resource_management.go [hunk 7]
```diff
@@ -316,6 +316,23 @@ func GetMultipleResources(context middleware.Context, dataStore db.DB, resourceS
 	currCond := policy.GetCurrentResourceCondition()
 
 	filter := FilterFromQueryParameter(resourceSchema, queryParameters)
+	if search_filter,ok := queryParameters["search_field"]; ok && resourceSchema.IsSubstringSeachEnabled(){
```
Comment (cierpuchaw) [inline server/resources/resource_management.go]: If we can enable substring searching on demand now, do we really need to have that also enabled in schemas?

File: server/resources/resource_management.go (hunk 3 above)
Comment (cierpuchaw) [reply server/resources/resource_management.go]: Yep, this is already called 2 lines below.

File: schema/schema.go [hunk 8]
```diff
@@ -278,6 +278,19 @@ func (schema *Schema) Init() error {
 	return nil
 }
 
+//Is substring search enabled for the schema
+func (schema *Schema) IsSubstringSeachEnabled() bool {
+	likeEnabled, ok := schema.Metadata["substr_search_enabled"]
+	if !ok {
+		return true
```
[the remark this answers is not in the document]
Comment (giyyanan) [reply schema/schema.go]: this one specifically for blocking sub string search on certain schema's like log tables

File: server/resources/resource_management.go (hunk 7 above)
Comment (giyyanan) [reply server/resources/resource_management.go]: earlier this flag was disable_substring_search...earlier reviews suggested no to stick to negative logic ..so changed it to enable_substring_search..probably the enable might be confusing..will try to find a different name
https://github.com/cloudwan/gohan/pull/731#discussion_r262837149

File: db/sql/sql_test.go [hunk 9]
```diff
@@ -411,6 +412,32 @@ var _ = Describe("Sql", func() {
...
+				number_search := search.NewSearchField("42")
+				filter := map[string]interface{}{"test_string": string_search, "test_number": number_search}
+
+				res, err := AddFilterToQuery(testSchema, query, filter, false)
+
+				Expect(err).ToNot(HaveOccurred())
+				resSql, param, err := res.ToSql()
+				Expect(err).ToNot(HaveOccurred())
+				checkAndStatement(resSql, "`test_string` LIKE ?", "`test_number` LIKE ?", param, []interface{}{ "%123%","%42%"})
+			})
```
Comment (morrisson) [inline db/sql/sql_test.go]: Let's add an unit test of SQL statement mixed with substr search enabled/disabled fields.

File: docs/schema.md [hunk 10]
```diff
@@ -654,6 +659,23 @@ HTTP Status Code: 200
     }
   }
 ```
+GET http://$GOHAN/[$namespace_prefix/]$prefix/$plural/$id&search_field=$id
```
Comment (morrisson) [inline docs/schema.md]: `/$id` is an example of GETting single resource, not multiple resources, so it's not a good place to explain substr match.

**Comment by morrisson** [comment]
@giyyanan  `go fmt` will help you to make your code well-formatted. I believe some plugins of auto `go fmt` should be available for your IDE.

File: server/server_test.go (hunk 6 above)
Comment (giyyanan) [reply server/server_test.go]: @cierpuchaw  exact matches would be the get default get queries ...wouldn't existing test case suffice for that... 

search_field..i could change it to search_on_field or search_column

File: db/search/search.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+const (
+	LIKE = "like"
+)
+
+type Search struct {
+	Value string
+}
+
+func NewSearchField(value string) Search {
+	searchValue := strings.Builder{}
```
Comment (p-kozlowski) [inline db/search/search.go]: this comment https://github.com/cloudwan/gohan/pull/731#discussion_r263075188 is still valid and needs to be resolved

File: db/sql/sql.go [hunk 12]
```diff
@@ -1274,6 +1282,12 @@ func addToFilter(s *schema.Schema, q sq.SelectBuilder, filter interface{}, join
 
 			// TODO: add other operators
 			value := filter["value"]
+			substr, ok := value.(search.Search)
+			if ok {
+				sqlizer = append(sqlizer, sq.Like{column: substr.Value})
```
Comment (p-kozlowski) [inline db/sql/sql.go]: there are not tests for it (and it looks like it breaks the API). please see how the "extended" API looks like and add tests for it [this](https://github.com/cloudwan/gohan/blob/master/db/sql/sql_test.go#L425) block

File: server/server_test.go [hunk 13]
```diff
@@ -1273,6 +1273,75 @@ var _ = Describe("Server package test", func() {
...
+
+		AfterEach(func() {
+			testURL("DELETE", getNetworkSingularURL("red"), adminTokenID, nil, http.StatusNoContent)
+			testURL("DELETE", getNetworkSingularURL("red1"), adminTokenID, nil, http.StatusNoContent)
+			testURL("DELETE", getNetworkSingularURL("red2"), adminTokenID, nil, http.StatusNoContent)
+			testURL("DELETE", getNetworkSingularURL("red3"), adminTokenID, nil, http.StatusNoContent)
+		})
+
+		DescribeTable("match all rows by substring search", func(query string) {
+			result := testURL("GET", networkPluralURL+query, adminTokenID, nil, http.StatusOK)
```
[the remark this answers is not in the document]
Comment (giyyanan) [reply server/server_test.go]: path seems to append a escape slash to an already escaped one..

networkURL is stored as string `http://localhost:19090` and path overtires it with `http:///localhost:19090` which causes test cases to fail...could use string builder

[the remark this answers is not in the document]
Comment (p-kozlowski) [reply server/server_test.go]: ah, I see. StringBuilder would be an overkill. let's keep it as-is

File: db/search/search.go (hunk 11 above)
Comment (giyyanan) [reply db/search/search.go]: In case we are escaping the characters '%','_','/' with slash or something, we need to make sure that when the query is executed the escape character is handled in the queries 
i.e for instance with maria DB
https://mariadb.com/kb/en/library/like/ you specify the escape character while executing like queries

otherwise `%`+'/'+`%` can be executed as `%`+'//'+`%` and might result in incorrect results if not handled.

Since we are using Masterminds/squirrel package i don't see a place where escape characters are handled while framing the query strings

https://github.com/Masterminds/squirrel/blob/55303df43ec0bf2b0ce3f9c3d8c3ce07022ef76b/expr.go#L169

Comment (zimnx) [reply db/search/search.go]: By default, escape character is set to backslash. 'squirrel' doesn't have to support changing it, we can go with backslash.
https://mariadb.com/kb/en/library/like/
> If you need to match the characters _ or %, you must escape them. By default, you can prefix the wildcard characters the backslash character \ to escape them. The backslash is used both to encode special characters like newlines when a string is parsed as well as to escape wildcards in a pattern after parsing. Thus, to match an actual backslash, you sometimes need to double-escape it as "\\\\".

Comment (giyyanan) [reply db/search/search.go]: yes...mariadb supports backslash...gohan is DB agnostic right..isn't that the reason why squirrel is used ....don't we need to handle for other databases as well.. I tested with sqlite and it yielded wrong results, when escaped with backlash..

Comment (zimnx) [reply db/search/search.go]: Well, in this case you can always implement you own Where matcher similar to existing sq.Like, which supports ESCAPE clause. All you have to do is to implement sq.Sqlizer interface on your custom struct.

Comment (p-kozlowski) [reply db/search/search.go]: implementing sq.Sqlizer sound a bit like an overkill for me. maybe I'm missing something, but can we implement [driver.Valuer](https://github.com/Masterminds/squirrel/blob/v1/expr.go#L183) instead and stay with regular sq.Like? eg.

```Go
func (s *Search) Value() (driver.Value, error) {
	buf := strings.Builder{}
	buf.WriteString("%")
	buf.WriteString(escape(s.value))
	buf.WriteString("%")
	return buf.String()
}
```

File: db/sql/sql.go [hunk 14]
```diff
@@ -1215,6 +1217,12 @@ func AddFilterToQuery(s *schema.Schema, q sq.SelectBuilder, filter map[string]in
 			column = quote(key)
 		}
 
+		substr, ok := value.(search.Search)
+		if ok {
+			q = q.Where(sq.Like{column: substr.Value})
```
Comment (p-kozlowski) [inline db/sql/sql.go]: if we go with implementing `driver.Valuer`, this will become simply
```Go
q = q.Where(sq.Like{column: substr})
```

File: examples/noauth/gohan.yaml [hunk 15]
```diff
@@ -6,10 +6,10 @@
 database:
     # yaml, json, sqlite3 and mysql supported
     # yaml and json db is for schema development purpose
-    type: "sqlite3"
+    type: "mysql"
     # connection string
     # it is file path for yaml, json and sqlite3 backend
-    connection: "./gohan.db"
+    connection: "chilli:passwd@tcp(127.0.0.1:3306)/gohan_test"
```
Comment (p-kozlowski) [inline examples/noauth/gohan.yaml]: private testing data?

File: schema/schema.go [hunk 16]
```diff
@@ -278,6 +278,15 @@ func (schema *Schema) Init() error {
 	return nil
 }
 
+//Is substring search enabled for the schema
+func (schema *Schema) IsSubstringSearchEnabled() bool {
+
```
Comment (p-kozlowski) [inline schema/schema.go]: redundant whitespace

File: server/server_test.go [hunk 17]
```diff
@@ -1273,6 +1273,76 @@ var _ = Describe("Server package test", func() {
...
+		DescribeTable("match and search combinations return two rows", func(query string) {
+			result := testURL("GET", networkPluralURL+query, adminTokenID, nil, http.StatusOK)
+			Expect(result).To(HaveKeyWithValue("networks", ConsistOf(
+				util.MatchAsJSON(getNetwork("red3", "blue")),
+				util.MatchAsJSON(getNetwork("red2", "blue")))))
+		},
+			Entry("match on id and search on description", "?id=networkred2&id=networkred3&description=red&search_field=description"),
+			Entry("match on tenant and search on id", "?id=red&tenant_id=blue&search_field=id"),
+			Entry("search on tenant", "?tenant_id=blue&search_field=tenant_id"),
+			Entry("non exclusive parameters","?id=re&search_field=id&tenant_id=bl&search_field=tenant_id"),
```
Comment (p-kozlowski) [inline server/server_test.go]: please run `goimports` (which is a superset of `go fmt`)

File: schema/schema.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (cierpuchaw) [reply schema/schema.go]: I don't get it. Why is substring searching **enabled** by default when schema does not enable or disable it explicitly? What's blocked here?

This function has had its contents changed completely but this property was retained.

File: examples/noauth/gohan.yaml (hunk 15 above)
Comment (giyyanan) [reply examples/noauth/gohan.yaml]: my bad..removed

File: schema/schema.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (giyyanan) [reply schema/schema.go]: the goal was to enable sub string search on all schema tables except for a few where the no of records run into millions i.e log tables..in such cases  we explicit block the search feature even if the API requests for it with the `search_field` parameter....

the naming is confusing but was changed as per previous comments...i would have preferred 'disable_substr_search'

File: db/sql/sql.go (hunk 12 above)
Comment (giyyanan) [reply db/sql/sql.go]: except for injecting the squirrel.Like in some of the already existing test cases couldn't find any need for new test cases...

File: db/search/search.go (hunk 11 above)
Comment (giyyanan) [reply db/search/search.go]: apart from this suggested change..i escaped characters and ran test with the server_test.go..still it fails in producing the correct results as i believe the test configuration is not mysql..so tests for the search feature fails with escaped chars..which brings to ..why aren't we testing the api's on a set of databases..rather than just sqlite3

Comment (zimnx) [reply db/search/search.go]: @p-kozlowski As far as I understand, sqlite doesn't support backslash escape character by default. To enable it, query must contain optional ESCAPE statement followed by escape character. In MariaDB, backslash is enabled by default. squirrel doesn't provide such query modifier, but exposes interace which could be implemented to add such field. 

@giyyanan i see that characters are escaped, and you wrote that some tests are failing. Those on CircleCI are passing. Does CI skips any tests?

Comment (giyyanan) [reply db/search/search.go]: I haven't checked in the code for unit test cases with special characters..

File: schema/schema.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (cierpuchaw) [reply schema/schema.go]: Got it!

File: server/server_test.go [hunk 18]
```diff
@@ -1273,6 +1273,139 @@ var _ = Describe("Server package test", func() {
...
+				util.MatchAsJSON(getNetwork("_apple", "apple")),
+				util.MatchAsJSON(getNetwork("\\ball", "ball")),
+				util.MatchAsJSON(getNetwork("%cats", "cats")),
+				util.MatchAsJSON(getNetwork("a\\x%y_z", "xyz")))))
+		},
+			Entry("empty query", ""),
+			Entry("common occuring character", "?id=a&search_field=id"),
+		)
+
+		PDescribeTable("search should return single row with _", func(query string) {
```
Comment (zimnx) [inline server/server_test.go]: Instead marking tests as Pending, it's better to implement [my previous comment](https://github.com/cloudwan/gohan/pull/734/files#r271553551) to support character escaping in both supported DB engines.

**Comment by morrisson** [comment]
@zimnx @p-kozlowski @cierpuchaw @michalzurawski @jkmar Please make a review again. Thanks.

File: db/sql/sql.go [hunk 19]
```diff
@@ -1299,3 +1315,48 @@ func (db *DB) SetMaxOpenConns(maxIdleConns int) {
...
+		case driver.Valuer:
+			if val, err = v.Value(); err != nil {
+				return
+			}
+		}
+
+		if val == nil {
+			err = fmt.Errorf("cannot use null with like operators")
+			return
+		} else {
```
Comment (cierpuchaw) [inline db/sql/sql.go]: This `else {` is not needed, we always return from the function when taking the first branch.

File: db/sql/sql.go [hunk 20]
```diff
@@ -1299,3 +1315,48 @@ func (db *DB) SetMaxOpenConns(maxIdleConns int) {
...
+			if isListType(val) {
+				err = fmt.Errorf("cannot use array or slice with like operators")
+				return
+			} else {
```
Comment (cierpuchaw) [inline db/sql/sql.go]: Same here.

File: db/sql/sql.go [hunk 21]
```diff
@@ -1299,3 +1315,48 @@ func (db *DB) SetMaxOpenConns(maxIdleConns int) {
 func safeMysqlContext(_ context.Context) context.Context {
 	return context.Background()
 }
+
+type Like map[string]interface{}
+
+func (lk Like) ToSql() (sql string, args []interface{}, err error) {
+	var (
+		exprs []string
```
Comment (cierpuchaw) [inline db/sql/sql.go]: `exprs` is either empty (in case of error) or has `len(lk)` elements, we can preallocate memory here.

File: db/sql/sql_test.go [hunk 22]
```diff
@@ -411,6 +412,57 @@ var _ = Describe("Sql", func() {
...
+
+				res, err := AddFilterToQuery(testSchema, query, filter, false)
+
+				Expect(err).ToNot(HaveOccurred())
+				resSql, param, err := res.ToSql()
+				Expect(err).ToNot(HaveOccurred())
+				expectedQuery = expectedQuery.Where(Like{"`test_string`": "%\\\\1\\%2\\_3\\_a\\\\b\\%c%"})
+				expectedSql, expectedParam, _ := expectedQuery.ToSql()
+				Expect(resSql).To(Equal(expectedSql))
+				Expect(param).To(Equal(expectedParam))
```
Comment (cierpuchaw) [inline db/sql/sql_test.go]: Above block of code is virtually identical to the one in the preceding test case. You can factor this out into a function.

File: db/sql/sql_test.go [hunk 23]
```diff
@@ -448,6 +501,31 @@ var _ = Describe("Sql", func() {
...
+				res, err := AddFilterToQuery(testSchema, query, filter, false)
+
+				Expect(err).ToNot(HaveOccurred())
+				resSql, param, err := res.ToSql()
+				Expect(err).ToNot(HaveOccurred())
+
+				expectedQuery = expectedQuery.Where(squirrel.Or{Like{"`test_string`": "%123%"}, Like{"`test_number`": "%42%"}})
+				expectedSql, expectedParam, _ := expectedQuery.ToSql()
+				Expect(resSql).To(Equal(expectedSql))
+				Expect(param).To(Equal(expectedParam))
```
Comment (cierpuchaw) [inline db/sql/sql_test.go]: Another candidate for using a helper function.

File: db/sql/sql_test.go [hunk 24]
```diff
@@ -473,6 +551,33 @@ var _ = Describe("Sql", func() {
...
+				res, err := AddFilterToQuery(testSchema, query, filter, false)
+
+				Expect(err).ToNot(HaveOccurred())
+				resSql, param, err := res.ToSql()
+				Expect(err).ToNot(HaveOccurred())
+
+				expectedQuery = expectedQuery.Where(squirrel.And{Like{"`test_string`": "%123%"}, Like{"`test_number`": "%42%"}})
+				expectedSql, expectedParam, _ := expectedQuery.ToSql()
+				Expect(resSql).To(Equal(expectedSql))
+				Expect(param).To(Equal(expectedParam))
```
Comment (cierpuchaw) [inline db/sql/sql_test.go]: Another candidate for using a helper function.

File: docs/schema.md [hunk 25]
```diff
@@ -145,6 +145,10 @@ Developers can make a schema as abstract schema specifying type=abstract. The de
 
   We don't sync this resource for sync backend when this option is true.
 
+- enable_substring_search (boolean)
+
+  can be used to block substring matches while fetching data in GET API request calls. By defult substring search i enabled for all schemas
```
Comment (cierpuchaw) [inline docs/schema.md]: Tiny typo: `search i enabled` -> `search is enabled`

File: db/sql/sql.go (hunk 21 above)
Comment (giyyanan) [reply db/sql/sql.go]: this just a copy of the function in squirrel..had to copy it here just for the ESCAPE part..intent of the function was to use as exactly as what squirrel library has and just inject ESCAPE for the queries..

Comment (michalzurawski) [reply db/sql/sql.go]: i guess Wojciech meant to change this line to `exprs = make([]string, 0, len(lk))` to avoid unnecessary allocation in for loop

File: db/sql/sql.go (hunk 19 above)
Comment (giyyanan) [reply db/sql/sql.go]: else where the ESCAPE injunction is

File: db/sql/sql.go (hunk 20 above)
Comment (giyyanan) [reply db/sql/sql.go]: else where the ESCAPE injunction is

File: db/sql/sql.go (hunk 19 above)
Comment (cierpuchaw) [reply db/sql/sql.go]: I'm not asking to remove the entire `else` block. Let me explain using an example.

We have a similar situation to the following:
```go
if cond1 {
  return
} else {
  if cond2 {
    return
  } else {
    doStuff()
  }
}
```

This is equivalent to:
```go
if cond1 {
  return
}

if cond2 {
  return
}

doStuff()
```

Doing this transformation reduces unneeded indentation, which with deep nesting, makes it hard to follow the code.

File: db/sql/sql_test.go [hunk 26]
```diff
@@ -422,6 +468,35 @@ var _ = Describe("Sql", func() {
 				checkAndStatement(resSql, "`test_string` = ?", "`test_number` = ?", param, []interface{}{42, "123"})
 			})
 		})
+		checkLikeDisjunctionAndConjunction := func(searchString string, numberString string, filterType string, convertedString string, convertedNumber string){
```
Comment (morrisson) [inline db/sql/sql_test.go]: I think we can consider one step more for generalized helper function and readable test cases not only for your added test cases. Does following function definition work for you?

```
checkTranslationFilterToWhereClause := func(filter map[string]interface{}, whereClause squirrel.Sqlizer) {
	res, err := AddFilterToQuery(testSchema, query, filter, false)

	Expect(err).ToNot(HaveOccurred())
	resSql, param, err := res.ToSql()
	Expect(err).ToNot(HaveOccurred())

	expectedQuery = expectedQuery.Where(whereClause)
	expectedSql, expectedParam, _ := expectedQuery.ToSql()
	Expect(resSql).To(Equal(expectedSql))
	Expect(param).To(Equal(expectedParam))
}
```

Maybe above helper function can be used by not only "Disjunction and conjunction" context but also "Basic select query".

File: db/sql/sql_test.go [hunk 27]
```diff
@@ -448,6 +523,9 @@ var _ = Describe("Sql", func() {
 				Expect(resSql).To(Equal(expectedSql))
 				Expect(param).To(Equal(expectedParam))
 			})
+			It("should create disjunction for more than one parameter with like queries", func() {
+				checkLikeDisjunctionAndConjunction("123","42","__or__","%123%","%42%")
```
Comment (morrisson) [inline db/sql/sql_test.go]: `go fmt` should make a space after each comma.

Comment (cierpuchaw) [reply db/sql/sql_test.go]: Tip: set an after-save-hook in your editor to run `gofmt`, or even better, `goimports`.

Comment (giyyanan) [reply db/sql/sql_test.go]: thanks @cierpuchaw

**Comment by morrisson** [comment]
Unfortunately, following files are changed unintentionally while you squash commits.

- .github/CODEOWNERS
- docs/extension.md
- server/api.go
- server/middleware/middleware.go

Please check your commits then fix to not include those files.

**Comment by morrisson** [comment]
@giyyanan Looks good. Thanks :)
